File: chap15_muscle/muscle_smooth.m

%  -------------------------------------------------------------------
%
% Hai-Murphy-Huxley model of smooth muscle. Use the method of
% characteristics to compute the distributions as functions of space and
% time. Use two different versions of the model.
%
%   For Chapter 15, Section 15.9.1 of
%   Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
%
%   Written by James Keener and James Sneyd.
%
%  -------------------------------------------------------------------

function muscle_smooth
close all
clear all
clc

par.f1 = 0.88; 
par.g1 = 0.21;
par.g2 = 4.4;
par.gL1 = 0.01;
par.gL2 = 0.2;
par.h = 1;
par.delx = 1.2;   % distance between binding sites
par.k1 = 0.35;
par.k2 = 0.1;
par.k5 = 0.35;
par.k6 = 0.1;

par.num = 350;      % number of space points
par.numt = 500;     % number of time outputs
par.tend = 20;       % final time

% calculate the shifts needed in the complicated model version
xlow = -6; xhigh = 6;
point_spacing = (xhigh-xlow)/par.num;
par.xshift = floor(par.delx/point_spacing);        % calculate how many points fit in length delx

% initial conditions. use the conservation law to determine Nmp
x0 = linspace(xlow,xhigh,par.num);
nm0 = ones(1,par.num);
nam0 = zeros(1,par.num);
namp0 = zeros(1,par.num);
y0 = [x0 nm0 nam0 namp0];
output_times = linspace(0,par.tend,par.numt);  % times to collect the output for plotting stuff


% First solve using the odes of the simplified model, one for each x point
[tout1,yout]=ode15s(@(t,y)derivs_chars(t,y,par),output_times,y0);
for i=1:par.numt
    x = yout(i,1:par.num);
    nam = yout(i,2*par.num+1:3*par.num);
    namp = yout(i,3*par.num+1:4*par.num);
    force1(i) = trapz(x,x.*(nam+namp));
end
figure(1)
plot(tout1,force1,'LineWidth',2)
hold on

% Now solve using the more complicated ODEs, with the detailed conservation law
[tout2,yout]=ode15s(@(t,y)derivs_chars_alt(t,y,par),output_times,y0);
for i=1:par.numt
    x = yout(i,1:par.num);
    nam = yout(i,2*par.num+1:3*par.num);
    namp = yout(i,3*par.num+1:4*par.num);
    force2(i) = trapz(x,x.*(nam+namp));
end
plot(tout2,force2,'LineWidth',2)

% tout1 = tout1';
% tout2 = tout2';
% save('temp.mat','tout1','force1','tout2','force2')

end % of main


%% RHS of odes for the method of characteristics. Simple version.
function out = derivs_chars(t,y,par)

x = y(1:par.num);
nm = y(par.num+1:2*par.num);
nam = y(2*par.num+1:3*par.num);
namp = y(3*par.num+1:4*par.num);
nmp = ones(par.num,1) - nm - nam - namp;

out(1:par.num) = -v(t);
out(1*par.num+1:2*par.num) = par.k2*nmp - par.k1*nm + gL(x,par).*nam  ;                     % Nm
out(2*par.num+1:3*par.num) = par.k6*namp - (par.k5+gL(x,par)).*nam  ;                       % Nam
out(3*par.num+1:4*par.num) = par.k5*nam + f(x,par).*nmp - (par.k6+g(x,par)).*namp;      % Namp
out = out';
end

%% RHS of odes for the method of characteristics. Complex version using the correct conservation law
function out = derivs_chars_alt(t,y,par)
x = y(1:par.num);
nm = y(par.num+1:2*par.num);
nam = y(2*par.num+1:3*par.num);
namp = y(3*par.num+1:4*par.num);

% Now do all the shifts of Nam. By hand is the easiest to follow, although it's wordy
s=par.xshift; % to save typing
nams1 = zeros(par.num,1);  nams1(s+1:end) = namp(1:end-s);  
nams2 = zeros(par.num,1);  nams2(2*s+1:end) = namp(1:end-2*s);
nams3 = zeros(par.num,1);  nams3(3*s+1:end) = namp(1:end-3*s);
namsm1 = zeros(par.num,1);  namsm1(1:end-s) = namp(s+1:end);  
namsm2 = zeros(par.num,1);  namsm2(1:end-2*s) = namp(2*s+1:end);
namsm3 = zeros(par.num,1);  namsm3(1:end-3*s) = namp(3*s+1:end);

% Finally, determine Nmp by using the conservation law. 
nmp = ones(par.num,1) - nm - namp - nam - nams1 - nams2 - nams3 - namsm1 - namsm2 - namsm3;

out(1:par.num) = -v(t);
out(par.num+1:2*par.num) = par.k2*nmp - par.k1*nm + gL(x,par).*nam + ...
                                             gL(x+par.delx,par).*nams1 + gL(x+2*par.delx,par).*nams2 + gL(x+3*par.delx,par).*nams3 + ...
                                             gL(x-par.delx,par).*namsm1 + gL(x-2*par.delx,par).*namsm2 + gL(x-3*par.delx,par).*namsm3;   % Nm
out(2*par.num+1:3*par.num) = par.k6*namp - (par.k5+gL(x,par)).*nam;                         % Nam
out(3*par.num+1:4*par.num) = par.k5*nam + f(x,par).*nmp - (par.k6+g(x,par)).*namp;      % Namp
out = out';
end

%% ------------
function out = f(x,par)
out = 0 + (x>0 & x<=1).*par.f1.*x;
end

function out = g(x,par)
out = (x<0).*par.g2 + (x>0).*par.g1.*x;
end

function out = gL(x,par)
out = (x<0).*par.gL2 + (x>0).*par.gL1.*x;
end

function out = v(t)
%out = 0.6; 
out = sin(2*t);
end

File: chap15_muscle/muscle_smooth.py

#   -------------------------------------------------------------------
# 
#  Hai-Murphy-Huxley model of smooth muscle. Use the method of
#  characteristics to compute the distributions as functions of space and
#  time. Use two different versions of the model.
# 
#    For Chapter 15, Section 15.8 of
#    Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
# 
#    Written by James Keener and James Sneyd.
# 
#   -------------------------------------------------------------------

import numpy as np
from scipy.integrate import solve_ivp
import matplotlib.pyplot as plt

def deriv(t,y):
    dd = np.zeros(4*num)
    x = y[0:num]
    nm = y[num:2*num]  
    nam = y[2*num:3*num]
    namp = y[3*num:4*num]
    nmp = np.ones(num) - nm - nam - namp
    
    dd[0:num] = -v(t)
    dd[num:2*num] = k2*nmp - k1*nm + ggL(x)*nam
    dd[2*num:3*num] = k6*namp - (k5+ggL(x))*nam
    dd[3*num:4*num] = k5*nam + ff(x)*nmp - (k6+gg(x))*namp
    return dd

def deriv_complex(t,y):
    dd = np.zeros(4*num)
    x = y[0:num]
    nm = y[num:2*num]  
    nam = y[2*num:3*num]
    namp = y[3*num:4*num]
    
    # Now do all the shifts of Nam. By hand is the easiest to follow, although it's wordy
    s=xshift # to save typing
    nams1 = np.zeros(num);  nams1[s:-1] = namp[0:-1-s]
    nams2 = np.zeros(num);  nams2[2*s:-1] = namp[0:-1-2*s];
    nams3 = np.zeros(num);  nams3[3*s:-1] = namp[0:-1-3*s];
    namsm1 = np.zeros(num);  namsm1[0:-1-s] = namp[s:-1];  
    namsm2 = np.zeros(num);  namsm2[0:-1-2*s] = namp[2*s:-1];
    namsm3 = np.zeros(num);  namsm3[0:-1-3*s] = namp[3*s:-1];
    nmp = np.ones(num) - nm - namp - nam - nams1 - nams2 - nams3 - namsm1 - namsm2 - namsm3;
    
    dd[0:num] = -v(t)
    dd[num:2*num] = k2*nmp - k1*nm + ggL(x)*nam + (
            ggL(x+delx)*nams1 + ggL(x+2*delx)*nams2 + ggL(x+3*delx)*nams3 +
            ggL(x-delx)*namsm1 + ggL(x-2*delx)*namsm2 + ggL(x-3*delx)*namsm3)
    dd[2*num:3*num] = k6*namp - (k5+ggL(x))*nam
    dd[3*num:4*num] = k5*nam + ff(x)*nmp - (k6+gg(x))*namp
    return dd

def f(x):
    return np.piecewise(x,[x<=0,0<=x<=1,x>1],[0,f1*x,0])
    
def g(x):
    return np.piecewise(x,[x<=0,0<=x],[g2,g1*x])

def gL(x):
    return np.piecewise(x,[x<=0,0<=x],[gL2,gL1*x])

def v(t):
    return np.sin(2*t)

# ----------------------------------------------------------------
ff = np.vectorize(f)
gg = np.vectorize(g)
ggL = np.vectorize(gL)

f1  = 0.88; 
g1  = 0.21;
g2  = 4.4;
gL1 = 0.01;
gL2 = 0.2;
h   = 1;
delx = 1.2;
k1  = 0.35;
k2  = 0.1;
k5  = 0.35;
k6  = 0.1;

num = 50;      # number of space points
numt = 100;     # number of time outputs
tend = 5;      # final time

# calculate the shifts needed in the complicated model version
xlow = -6; xhigh = 6;
point_spacing = (xhigh-xlow)/num;
xshift = round(delx/point_spacing);        # calculate how many points fit in length delx

# Initial conditions
x0 = np.linspace(xlow,xhigh,num);
nm0 = np.zeros(num)
nam0 = np.zeros(num)
namp0 = np.zeros(num)
y0 = np.concatenate((x0,nm0,nam0,namp0))  # the final ode is for the spring length
tout = np.linspace(0,tend,numt)

# solve the odes
soln = solve_ivp(deriv,[0, tend],y0,method='Radau',t_eval=tout,rtol=10e-6,atol=10e-9)

# calculate and plot the force
plt.figure(1)
p = np.zeros(numt)
for i in range(0,numt):
    x = soln.y[0:num,i]
    nam = soln.y[2*num:3*num,i]
    namp = soln.y[3*num:4*num,i]
    p[i] = np.trapz( x*(nam+namp),x )
plt.plot(soln.t,p)

# solve the more complex odes
soln = solve_ivp(deriv_complex,[0, tend],y0,method='Radau',t_eval=tout,rtol=10e-6,atol=10e-9)

# calculate and plot the force
plt.figure(1)
p = np.zeros(numt)
for i in range(0,numt):
    x = soln.y[0:num,i]
    nam = soln.y[2*num:3*num,i]
    namp = soln.y[3*num:4*num,i]
    p[i] = np.trapz( x*(nam+namp),x )
plt.plot(soln.t,p)
